zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
+            let orchard_read_task = std::thread::spawn(move || {
+                for (height, tree) in
+                    db.orchard_tree_by_height_range(orchard_height..initial_tip_height)
+                {
+                    // Breaking from this loop and dropping the orchard_tree channel
+                    // will cause the orchard compare and delete tasks to finish.
+                    if should_cancel_flag.load(atomic::Ordering::Relaxed) {
+                        break;
+                    }
+
+                    if let Err(error) = orchard_tree_tx.send((height, tree)) {
+                        warn!(?error, "unexpected send error")
+                    }
+                }
+            });
+
+            // Create an unbounded channel for duplicate orchard note commitment tree heights
+            let (unique_orchard_tree_height_tx, unique_orchard_tree_height_rx) = mpsc::channel();
+
+            // Set up task for reading orchard note commitment trees
+            let orchard_compare_task = std::thread::spawn(move || {
+                while let Ok((height, tree)) = orchard_tree_rx.recv() {
+                    let tree = Some(tree);
+                    if prev_orchard_tree != tree {
+                        if let Err(error) = unique_orchard_tree_height_tx.send(height) {
+                            warn!(?error, "unexpected send error")
+                        }
+
+                        prev_orchard_tree = tree;
+                    }
+                }
+            });
+
+            // Set up task for deleting orchard note commitment trees
+            let db = upgrade_db;
+            let orchard_delete_task = std::thread::spawn(move || {
+                let mut delete_from = orchard_height.next();
+                while let Ok(delete_to) = unique_orchard_tree_height_rx.recv() {
+                    let num_entries = delete_to - delete_from;
+                    if num_entries > 0 {
+                        let mut batch = DiskWriteBatch::new();
+
+                        if num_entries == 1 {
+                            batch.delete_orchard_tree(&db, &delete_from);
+                        } else {
+                            batch.delete_range_orchard_tree(&db, &delete_from, &delete_to);
+                        }
+
+                        db.write_batch(batch).expect(
+                            "Deleting Orchard note commitment trees should always succeed.",
+                        );
+                    }
+
+                    delete_from = delete_to.next();
+                }
+            });
+
+            for task in [
+                sapling_read_task,
+                sapling_compare_task,
+                sapling_delete_task,
+                orchard_read_task,
+                orchard_compare_task,
+                orchard_delete_task,
+            ] {
+                if let Err(error) = task.join() {
+                    warn!(?error, "unexpected join error")
```

Comment:
Optional:

#### Related Logging Bug

Threads need to be explicitly set up with the logging spans, otherwise their logs can be missing important diagnostic info.

#### Related Performance Question

Are these threads actually more efficient than running all the code in the upgrade thread? Does the difference matter?

Splitting code into separate threads can actually slow it down due to lock contention, context switching overhead, and missed compiler optimisations. (Code that's split across threads can't be optimised using assumptions from code in another thread.)

Reply:
[cd4be3e](https://github.com/ZcashFoundation/zebra/pull/7379/commits/cd4be3e2f6829994c7b7c2989f5b39fd06141ecf) removes the threads.